[FIX] web_editor: fix issues related to pasting and links

cammarosano opened the pull request:
This PR addresses 3 issues:

~1. Selection restore issues after pasting:
After pasting content like "text www.someurl.com more_text" on a non-collapsed selection followed by UNDO, the selection was not restored to the original one. This PR fixes it.~ 

~2. Pasting content in a empty link or in a selection spanning the whole content of a link:
In order to prevent inadvertently pasting content **inside** a link rather than replacing it, this PR removes the link when pasting content inside an empty link or when all of its text content is selected, letting the pasted content replace it.~

**(these issues are now addressed by PR https://github.com/odoo/odoo/pull/108755)**

3. Bad use of Node.textContent:
Misuse of this method was causing the following bugs:
    - parsing the wrong URL when updating a link's href based on its label
    - changing an elements content on cleanForSave if a ZWS inline element contained children element other than text nodes.
This PR replaces the use of textContent for safer alternatives.  

Task-3081254.

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/108521).

dmo-odoo commented on addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:
```diff
@@ -3173,10 +3194,12 @@ export class OdooEditor extends EventTarget {
+            this.addImagesFiles(files).then(html => this._applyCommand('insertHTML', this._prepareClipboardData(html)));
         } else if (clipboardHtml && targetSupportsHtmlContent) {
-            this.execCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
+            this._applyCommand('insertHTML', this._prepareClipboardData(clipboardHtml));
```
Shouldn't that have been changed in the previous commit ?

dmo-odoo replied:
My bad, I thought it was a line that ended up in the wrong commit.

dmo-odoo commented on addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:
```diff
@@ -3173,10 +3194,12 @@ export class OdooEditor extends EventTarget {
         const targetSupportsHtmlContent = isHtmlContentSupported(sel.anchorNode);
         const clipboardHtml = ev.clipboardData.getData('text/html');
+        // if pasting on a link whose label is fully selected, replace it by incoming content
+        this._removeFullySelectedLink();
```
I think the two utils function (both used only once) only add more boilerplate to an issue that would otherwise be quite simple. Therefore, I suggest :
```js
// Replace entire link if its label is fully selected.
const selection = this.document.getSelection();
const link = closestElement(selection.anchorNode, 'a', true);
if (link && selection.toString() === const innerText = element.innerText.replace(/\u200B/g, '')) {. 
    const start = leftPos(link);
    // Disable isolatated link.
    if (this._fixLinkMutatedElements && this._fixLinkMutatedElements.link === link) {
        this.resetContenteditableLink();
        this._activateContenteditable();
    }
    link.remove();
    setSelection(...start, ...start, false);
}
```
Besides, it is not clear to me why we need to disable the isolated link. Why is that ? If it is a valid reason then I think it would greatly benefit from a comment.

cammarosano replied:
`resetContenteditableLink` clears up `this.fixLinkMutatedElements`. I could do a `delete this.fixLinkMutatedElements` instead, if you prefer.

`activateContenteditable` is necessary to be able to place the cursor somewhere outside the link that is about to be deleted (and to continue editing the document).

dmo-odoo replied:
Ok so the comment could be: `// Exit link isolation since we're removing the link and editing outside of it.`

cammarosano replied:
Pushed updated code with suggested changes.

dmo-odoo commented on addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:
```diff
@@ -2442,3 +2442,16 @@ export function selectionMatchesInnerText(element) {
+/**
+ * Removes all ZeroWidthSpace characters from a node (if text) or
+ * its descendants.
+ * It does not remove nodes.
```
I suggest (break at 80 chars):
Remove zero-width spaces from the provided node and its descendants.

dmo-odoo commented on addons/web_editor/static/lib/odoo-editor/src/utils/utils.js:
```diff
@@ -2442,3 +2442,16 @@ export function selectionMatchesInnerText(element) {
+ *
+ * @param {Node} node
+ */
+export function removeZWSChars(node) {
```
I suggest renaming this function to `removeZWS` to be consistent with `isZWS`.

cammarosano replied:
`isZWS` informs whether a **node** contains only a ZWS. Thus, is node "ZWS"? True or false.
`removeZWS` would make me think the function removes such node. But it does not, it removes ZeroWidthSpace characters in a node's content. The node is kept. It is meant to remove ZWS from nodes/elements that have ZWS and other stuff.
I agree that `removeZWS` is more consistent with that other function's name, but it could be deceiving.

cammarosano commented on addons/web_editor/static/lib/odoo-editor/src/OdooEditor.js:
```diff
@@ -2819,16 +2821,16 @@ export class OdooEditor extends EventTarget {
-            if (emptyElement.textContent.length === 1 && emptyElement.textContent.includes('\u200B')) {
-                if (emptyElement.classList.length > 0) {
-                    emptyElement.removeAttribute('oe-zws-empty-inline');
+        for (const el of element.querySelectorAll('[oe-zws-empty-inline]')) {
```
I would agree with `cleanZWS`.
But if avoid creating new functions is a must, I can just transform the snippet below into a one-liner and use it twice:
```
 [el, ...descendants(el)]
            .filter(node => node.nodeType === Node.TEXT_NODE)
            .forEach(node => node.nodeValue = node.nodeValue.replace(/\u200B/g, ''));
```       
(Actually it already is, it was just spread in 3 lines for readability).
But it could be rewritten as 
`[element, ...descendants(element)].forEach(node => node.nodeValue && node.nodeValue.replace(/\u200B/g, ''));`
I consider it to be a bit less readable, but it it's idiomatic in our code base and does save lines.

About your suggestion about merging the two querySelectorAll in one:
FA-elements would satisfy the first `if` condition (`isZWS`) and never reach that outer `else` where the ZWS cleaning happen. So I'd rather keep them separated (it's also more readable, right?).

dmo-odoo replied:
Avoiding creating new functions is not a must per se. If we agree on the name `cleanZWS` then I think it is a valuable util function and we should therefore keep it. 👍

cammarosano commented:
I removed the first commit, that addressed selection restore issues after pasting an undoing.
It turns out other selection restore issues were found, among other issues. They are all addressed by PR https://github.com/odoo/odoo/pull/108755.

cammarosano commented on addons/web_editor/static/lib/odoo-editor/src/utils/sanitize.js:
```diff
@@ -192,16 +191,17 @@ class Sanitize {
-                    node.setAttribute('href', urlInfo.url);
+                const linkLabel = node.innerText;
+                const urlInfo = getUrlsInfosInString(linkLabel);
+                if (urlInfo.length && urlInfo[0].label === linkLabel && !node.href.startsWith('mailto:')) {
```
When rebasing to force-push, I had to incorporate the changes introduced by this commit:
https://github.com/odoo/odoo/pull/107603/commits/1d8ca674c31ddd60680f889cf25d6c52740b54a0

Unless I'm missing something here, the commit does not do what it says it does: it does not "set link label as new URL". It simply prevents the href update mechanism.
It does solve the problem of breaking mailto links with the addition of "https://", but does not update the link if the email address is changed.

Such suitable link update (working for both https and mailto links) is the subject of task https://www.odoo.com/web#id=3102761&menu_id=4720&cids=1&action=333&active_id=1695&model=project.task&view_type=form

Should I take on this task in this PR? Or should I do it on a new PR targeting master?

dmo-odoo replied:
You don't _have_ to do it in this PR but if you think it would be easier to do so (I suspect it would in terms of future conflicts) then of course you can do it ! :)

cammarosano replied:
A proper link update supporting `http`, `mailto` and `tel` links is being done at https://github.com/odoo/odoo/pull/111521 targeting master. I'd rather leave this PR as it is, and later back-port the link update part for `mailto` if necessary.

cammarosano commented:
> I removed the first commit, that addressed selection restore issues after pasting an undoing. It turns out other selection restore issues were found, among other issues. They are all addressed by PR #108755.

I also removed the second commit, that removed a link when pasting over its whole text content, and put into PR #108755. This way, the security flag triggered by 'insertHTML' command in this commit can be addressed in that PR, which also has the same security issue flagged.

dmo-odoo commented:
@robodoo r+

robodoo commented:
@cammarosano @dmo-odoo staging failed: ci/runbot on e1157dbeb33c0e5ce7e7052240dc07d692c3049d (view more at https://runbot.odoo.com/runbot/build/23736212)

dmo-odoo commented:
@robodoo retry
